### DISCUSSION TITLE
OSC Input for Camera Control

### DISCUSSION THREAD

**Discussion by kobi32768** [vrchat-community/osc#50] [Ideas]
EN:
I think that it would be nice if all camera settings that can be changed from within the game (e.g. zoom and focus settings) could be changed via OSC.
I also think it would be better if we could use OSC communication as a trigger to take a picture, or if possible, a timer with a specified number of seconds.

JP:
私はゲーム内から変更できる全てのカメラの設定(例えば、ズームやフォーカスの設定)を OSC 経由で変更できたらいいのにと考えました。
また、OSC通信をトリガーに撮影したり、可能であれば秒数を指定したタイマーなどが使えたらより良くなるのではないかと思います。

**Comment by knah** [comment]
As an addition to this, allowing to move and rotate the camera via OSC also would make sense. This would allow animating camera movement, making recording of "cinematic" takes possible.

**Comment by anonmon** [comment]
Along with controls for the camera, being able to control the X/Y/Z position and rotation of the camera, in both either player and world space would be extremely useful. Between being able to make discreet controls for general movement, virtual rail and track shots, setting specific positions in space to be able to take consistent pictures in a world over multiple sessions, all sorts of options open up for photography and filming.
And I'd say more importantly, by also taking in a VR tracker's position/rotation with hypothetical OSC injector software to then translate to the VRC camera, that would open up SO much,
- Having a physical thing to hold to manipulate the VRC camera (Including IRL selfie sticks for the virtual camera if inclined)
- Sticking a small display onto the same platform as the tracker, meaning being able to view/record the virtual world as either a 2nd person in the playspace or just generally without having to be in VR.
- A small display + a IRL camera on the same platform as the tracker, mixed reality filming/photography with others virtually in VRChat being set into your IRL playspace would be possible.

**Comment by EIA485** [comment]
would also like to change all of the camera's setting via osc,
focus distance/focus point, aperture, fov. etc

**Comment by Macil** [comment]
Controlling the position and settings of a desktop camera through OSC while a player is in VR would useful. This could be used to:
- allow a player (using a separate OSC control or program while playing) with a camera facing themselves from a distance to rotate the camera to a new angle around them without having to move to reach the camera.
- allow a player to have the camera go on a scripted track around them or elsewhere.
- allow a player to let someone else who is watching the desktop view to adjust the camera. This could be a person sitting at the desktop watching their friend in VR, or it could be a remote person helping manage the recording of a streamer.
- allow a streamer to have twitch redeems that do something with the desktop camera, like move it to a funny angle or activate fisheye lens, etc.

To do these ideas, it would be useful to be able to have OSC inputs to control the position, angle, position mode, FOV, focal distance, focal mode, whether the camera is on or off, aim-at-player mode, and position anchor mode.

A little separately from the OSC part, it would also be useful to have a position anchor mode where the camera was anchored to the player's playspace but ignoring rotation, so the camera would stay aligned with the world and not jarringly jump and rotate when the player uses the controller to turn, but it would still shift its position to follow the player around when they use the controller to move.

**Comment by songminwoo1** [comment]
https://booth.pm/ko/items/3663253
maybe you can use this!

**Comment by kobi32768** [comment]
Thank you.
I knew that. But that is for VRCLens.
I want to control VRChat Official Camera.

**Comment by NjnaGrimsdottir** [comment]
I'm curious how this process works. I see there's been no comments or information on this since March... does that mean this isn't going to happen?  I was quite excited by the potential of this feature.

**Comment by RicardoEscobar** [comment]
IS there any progress on this? I'm looing for a way  to automate taking pictures every 5 seconds or so? I want to automate picture taking in a loop until I decide it to stop.

I tried with python and positioning the mouse over the button in a script to click the button every 5 seconds but it filed, it seems like VRChat blocks external programs pressing keys on keyboard and mouse inputs, but that's a guess I'm not sure.

**Reply by EIA485** [reply]
ah yep... the anti cheat may throw a fit about emulating user input like that.
